internal/pubsub/awspubsub/awspubsub.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package awspubsub provides an implementation of pubsub that uses AWS
+// SNS (Simple Notification Service) and SQS (Simple Queueing Service).
+//
+// It exposes the following types for As:
+// Topic: *sns.SNS
+// Subscription: *sqs.SQS
+package awspubsub
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"gocloud.dev/internal/pubsub"
+	"gocloud.dev/internal/pubsub/driver"
+)
+
+type topic struct {
+	mu     sync.Mutex
```

Comment:
You may not need the mutex—nothing changes, right?

Reply:
I don't know. Running with -race seems ok without this mutex, but maybe it's just because of insufficient test coverage.

Reply:
Oh, never mind. Their docs say the SNS clients are safe to use concurrently. https://docs.aws.amazon.com/sdk-for-go/api/service/sns/